src/flu/Flu_Tree_Browser.cpp

```diff
@@ -2983,7 +2983,7 @@ Flu_Tree_Browser::Node* Flu_Tree_Browser :: add_leaf( const char* path, const ch
 
 unsigned long Flu_Tree_Browser :: remove( const char *fullpath )
 {
-  return( (unsigned long)root.modify( fullpath, Node::REMOVE, rdata ) );
+  return reinterpret_cast<uintptr_t>(root.modify( fullpath, Node::REMOVE, rdata ) );
```

Comment:
This method returns a long, but the return is cast to uintptr.  Seems wrong to me.

Reply:
This is a 64-bit issue; the code is using a pointer as if it is a number (eg: pointer to unsigned long).  This is no longer valid in general (ptr size not matching size of a long).  The reinterpret cast changes the pointer value into a number (doing whatever is necessary for the platform at hand; uintptr is the magic type for this), then C++ auto-converts from that number to an unsigned long.

Reply:
I understand the idea, but the code not casts long to uintptr and then -- if I understand it correctly -- the return value is cast to long again because the method returns long.  In the best case, this seems to be a noop.

On windows, size(long) < size(ptr), so there is a problem with addressability somewhere already and that underlying problem should be rather fixed.

In addition, this code does not compile on my debian with gcc 10 because `cstdint` is not included.

Reply:
Did some more digging.  The issue is that Node::modify returns a Node*, but the value is not always a pointer.  You can see this, for example around line 3353 the code returns the actual pointer to a Node, while in the same function, around line 3506 it returns the ID (an int) cast to a pointer.  The remove() function is documented in the header file to "return the unique ID of the node removed or 0 if none was found".  So, the reinterpret cast in this context is correct (pointer is converted to a number, zero is left alone).

Reply:
The point I was trying to make is that this method returns a long (line 2984) and therefore the cast inside the method is pointless because it is then cast to long again upon return

Reply:
Sorry its late here, will  reply with more sleep

Reply:
I couldn't remember why I used the reinterpret cast in the first place, so I checked in a hard-cast which compiles fine on my Mac.  Then I remembered that this is the Windows branch.. so back to Windows and MINGW... This is the error it is giving me with a hard-cast.  Reinterpret-cast compiles cleanly on all platforms for me (Mac, Ubuntu, Windows).

gui/../flu/Flu_Tree_Browser.h:1075:12: error: cast from 'Flu_Tree_Browser::Node*' to 'long unsigned int' loses precision [-fpermissive]
 1075 |     return (unsigned long)ret;
      |            ^~~~~~~~~~~~~~~~~~

Reply:
Let me rephrase my critique: It does not matter what kind of casting is going on inside this method because the method itself returns a value of type double.  All the casts are doing is quieting the compiler.

Under Linux and Mac, long and pointer have the same size, so all the casts are noops. Under Win64, a long is only 32 bit but a pointer is 64 bit. The method returns a long, i.e. 32 bit. Converting a 32 bit double to uintptr (i.e. 64 bit) and then have the compiler add an implicit cast to double again (i.e. 32 bit) does not do anything meaningful as well.

The problem here are the return types that would need to be changed.

We can still merge this as it creates no new harm, but in that case we should have a TODO somewhere.